radixdlt/src/main/java/com/radixdlt/utils/Panicker.java

```diff
@@ -0,0 +1,159 @@
+/*
+ * (C) Copyright 2020 Radix DLT Ltd
+ *
+ * Radix DLT Ltd licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except in
+ * compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied.  See the License for the specific
+ * language governing permissions and limitations under the License.
+ */
+
+package com.radixdlt.utils;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.message.FormattedMessage;
+
+import com.google.common.annotations.VisibleForTesting;
+
+/**
+ * Panic methods for exiting system abruptly when invariants fail.
+ */
+public final class Panicker {
+
+	/**
+	 * We exit the application with this status code when we discover
+	 * an logic or data corruption error that cannot be recovered
+	 * without some intervention.
+	 * <p>
+	 * Note that the value here is {@code 128 + SIGABRT}, which is the
+	 * usual {@code libc} exit code when calling {@code abort()}.
+	 */
+	public static final int PANIC_EXIT_STATUS = 128 + 6;
+
+	// Allow panicking without exiting for tests
+	private static final AtomicBoolean exitOnPanic = new AtomicBoolean(true);
+
+	private Panicker() {
+		throw new IllegalStateException("Can't construct");
+	}
+
+	@VisibleForTesting
+	static final class PanicException extends RuntimeException {
+		private PanicException() {
+			super("panic");
+		}
+	}
+
+	/**
+	 * An error thrown by {@code Panicker.panic(...)} methods when a
+	 * panic occurs and the {@code exitOnPanic} flag is set to false.
+	 *
+	 * @see Panicker#setExitOnPanic(boolean)
+	 */
+	@VisibleForTesting
+	public static final class PanicError extends Error {
```

Comment:
Im a bit doubtful on use of this Error. It seems to me that any application error should at worst always be catchable.

Reply:
Anything that is a Throwable is catchable?  See also unit tests for Panicker.